### ISSUE TITLE
Add a way to read all enabled extensions across all extension hosts

### ISSUE THREAD

**Issue by alexdima** [microsoft/vscode#145307]
`vscode.extensions.all` or `vscode.extensions.getExtension` will only return extensions from the current extension host. The root cause for this is that the return type `IExtension<T>` exposes the `.exports` property, which represents the extension's exported API. Since extensions can export synchronous APIs, we can't automatically proxy it to make it available to other extension hosts. 

Also, making `vscode.extensions.getExtension` return an extension running on a different extension host, has the potential to cause breakage in existing extensions.

We could add a new function to be able to get to extensions that are running on other extension hosts.

This came up before e.g. in #82633

cc @jrieken

**Comment by aeschli** [comment]
The main motivation for me is to get access to the manifests of all extensions across extension hosts to find contributions such as `jsonValidation`. 

The same issue came up with the new anycode contributions and the `htmlLanguage` contribution were exploring in #146731.

**Comment by jrieken** [comment]
* The simple API change is another overload of the `vscode.extensions.getExtension`-function, like 
```ts
export function getExtension(extensionId: string, includeAllExtensionHost: boolean): Extension<unknown> | undefined;
```

* Also add another property for the [Extension](https://github.com/microsoft/vscode/blob/b261d64b68c6c6a5d261f7088d59dedba51a04d3/src/vscode-dts/vscode.d.ts#L6503)-type that allows to tell "same-ext-host" extensions apart from "foreign-ext-host" extension. Maybe a boolean like `Extension.isDifferentThanThisExtensionHost` or some kind of extension host identifier which you can compare with something like `vscode.env.thisExtensionHostIdentifier`
* Last, a good way to iterate over really all extension descriptions. Something like `vscode.extensions.allAll` to complement the `all` property. Or we review existing usages to decide for the breaking semantic change...

**Comment by DanTup** [comment]
@alexdima I noticed in the change above, `exports` will always be `undefined` for extensions in other hosts. Will there be any way for extensions in different extension hosts to communicate (for example can they call each others registered commands)?

**Comment by jrieken** [comment]
Yes, extension across different extension host can use commands to communicate with each other. However, arguments and returned values must be primitive or JSON-able

**Comment by DanTup** [comment]
@jrieken perfect, thanks!

**Comment by slackner** [comment]
Thanks for taking care of this issue!

We also need support for this feature for our [SemanticDiff](https://marketplace.visualstudio.com/items?itemName=semanticdiff.semanticdiff) plugin to properly support Dev Containers. To implement syntax highlighting, we currently enumerate grammars and themes contributed via extensions. When running in a dev container, however, in particular the `vscode.*` builtin extensions are not available.

Everything works correctly when using the `allAcrossExtensionHosts` property instead of `all`, so looking forward to get this proposal accepted.

**Comment by samdenty** [comment]
I'm interested in when support comes for extension communication via the exports. We ran into this limitation with [codeflow](https://stackblitz.com/codeflow)

**Comment by Laffery** [comment]
Perfect! When will this API be stable?

**Comment by zWingz** [comment]
When will this API be stable?

**Comment by a-stewart** [comment]
This api has had the tag `api-finalization` since April 2022 - is there anything we can do to help with the finalisation of the API?

In terms of feedback for the API, we have been using it with a few of our extensions, and the API itself seems fine, but there are a few comments on documentation.

Here is what is currently in vscode.d.ts

```ts
	export namespace extensions {

		/**
		 * Get an extension by its full identifier in the form of: `publisher.name`.
		 *
		 * @param extensionId An extension identifier.
		 * @returns An extension or `undefined`.
		 */
		export function getExtension<T = any>(extensionId: string): Extension<T> | undefined;

		/**
		 * All extensions currently known to the system.
		 */
		export const all: readonly Extension<any>[];

		/**
		 * An event which fires when `extensions.all` changes. This can happen when extensions are
		 * installed, uninstalled, enabled or disabled.
		 */
		export const onDidChange: Event<void>;
	}
```

`.getExtension()` doesn't make any mention that an installed extension might not be returned if it is in a different extension host.

`.all` says: "All extensions currently known to the system.", which is incorrect, even with the proposed api, since (according to this FR) you need to call `.allAcrossExtensionHosts`, in order to get the extensions on other hosts.

The name `.all` is also not ideal, but I do understand why you can't change that for backwards compatibility. (Although in that case, allowing multiple extension hosts initially was probably a breaking change too?)

It is also not clear if `.all` will include extensions which do not run in an extension host, eg language extensions. My understanding is that it will, but I think the docstring for this should explicitly state exactly what `.all` includes.

The event `onDidChange` may be a bit ambiguous once this API is finalised, will it fire for changes in other hosts?

---

Summary: Even if we don't finalise the API, I think the docstrings currently in vscode.d.ts should be updated, as they don't make it clear what will be returned by the APIs.

**Comment by a-stewart** [comment]
Thanks for adding this into the next milestone. I thought we should just follow up with some more findings as we are using separate extension hosts a lot more now and have seem a few differnt examples.

This is all just from our experiences, and there are likely additional cases that we haven't come across.

---

`.getExtension` and `.all` not returning extensions from other hosts has caused us issues. In particular we often have extensions wanting to read the manifest of another extension. Or we have an extension wanting to generate a list of all installed extensions. In both these cases we had to modify them to use the proposed API.

---

If an extension does not want to use `.exports`, we have not managed to come across any scenario where it would be beneficial not to return extensions from other hosts.

Even in the event that an extension does want to use the `exports` of another extension, there would not be any significant regression by changing the default behaviour to return the extension regardless of host. The change would be to go from it not working because the extension is not found, to not working because it is found but doesn't export anything.

In both situations the fix is the same, ensuring both extensions live in the same host.

---

The risk is in making a non backwards compatible change to the API. Currently you can do the following to programatically handle the extension not existing or being in a different host if you need access to `exports`.

```ts
const extension = vscode.extensions.getExtension<Foo>('myExtension');
if (extension) {
  extension.exports.bar();
}
```

Whereas if getExtension were to return an extension from another host, that code above would error as the check would not verify that the extension returns anything.
(But then there never actually was any guarantee in the first place, since the extension author manually specifies the type anyway.)

The question is how often is extension A going to call extension B, knowing with certainty that it exports something and of what type, without A listing B as a dependency, which pins them into the same host as far as I understand?

---

I think whilst this is also true for `.all` - it is very unlikely that anyone will be trying to access the exports of an extension from `.all`, especially without checking that they exist, so I think it would be safe to update `.all` to return all extensions. I think `onDidChange` should reflect what `.all` is.


I also think that a breaking change to the API is unavoidable.

vscode.extensions.all previously returned all extensions installed on that instance of VS Code. And the docstring says otherwise. Having `.all` not return extensions in other hosts is already a breaking change to the API, and I would argue is a bigger break, than having it return extensions in other hosts which have `exports = undefined`.

---

I would also consider modifying the proposal for getExtension to overload based on whether other hosts are requested, for example:

```ts
export function getExtension<T = any>(extensionId: string, includeDifferentExtensionHosts?: false): Extension<T> | undefined;
export function getExtension<T = any>(extensionId: string, includeDifferentExtensionHosts: true): Extension<T|undefined> | undefined;
export function getExtension<T = any>(extensionId: string, includeDifferentExtensionHosts: boolean|undefined): Extension<T> | Extension<T|undefined> | undefined;
```

This way if you call, `getExtension<Foo>('my.extension', /*allHosts*/ true);` you would get a `vscode.Extension<Foo|undefined>` making it clear that exports may not exist. But if you call `getExtension<Foo>('my.extension', /*allHosts*/ false);` then that would return the more previous `Extension<T>`. (You could argue whether `true` or `false` should be default.)

---

tl;dr - there is a breaking change the API whatever you do, but I think with `.all`, having it do what it says it does currently is the smaller break. With `.getExtension` you could easily argue both ways but it is less important since the ability to add a parameter mitigates this mostly.

**Comment by jrieken** [comment]
Better name suggestion from @Tyriar `allOnAnyExtensionHost `

**Comment by a-stewart** [comment]
With regards to `all` or `allOnAnyExtensionHost` - is there an example you can think of where, you might call `all` and not want it to include all extensions across all hosts?

**Comment by jrieken** [comment]
> you might call all and not want it to include all extensions across all hosts?

I don't know but we don't want to alter/break existing behaviour because there is a difference in what we return here. "Remote" extension don't have the API object set and today, with `all`, you never have that case

**Comment by yCodeTech** [comment]
CC: @jrieken @alexdima 
Can we please get this proposed API in the next update 🙏 This is a greatly needed API change!

 I develop [Automatic Comment Blocks](https://github.com/yCodeTech/auto-comment-blocks) and I use the `extensions.all` API to get all extensions and get all the installed language configs so my extension has automatic language support. All works as intended. But it's recently come to my attention, that it breaks on WSL2 because it can't get *`all`* extensions across hosts (https://github.com/yCodeTech/auto-comment-blocks/issues/6).

Ideally the `extensions.all` should get *`all`* extensions across hosts by default, but I also welcome this new proposed `extensions.allAcrossExtensionHosts` API, which I can see is already in the sourcecode: https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/common/extHost.api.impl.ts#L517C4-L526C6

I (and many other extension developers) would **really** love and appreciate it if this proposed API was to be finalised and make it's way into the next update.
